keylime/registrar_common.py

```diff
@@ -291,20 +292,73 @@ def do_POST(self) -> None:
+            idevid_cert = ""
+            iak_cert = ""
             iak_attest = b""
             iak_sign = b""
-            # Check if IDevID and IAK are received
-            # If not it is reported with logger info, but this behaviour will change when configuration requirements are added
-            idevid_received = False
-            if "iak_tpm" in json_body:
-                iak_tpm = json_body["iak_tpm"]
-                idevid_tpm = json_body["idevid_tpm"]
-                iak_attest = base64.b64decode(json_body["iak_attest"])
-                iak_sign = base64.b64decode(json_body["iak_sign"])
-                logger.info("IDevID and IAK received")
-                idevid_received = True
-            else:
-                logger.info("No IDevID/IAK received")
+            idevid_required = config.getboolean("registrar", "require_idevid_iak")
+            # If IDevID and IAK are required in the config:
+            #  -Check if IDevID and IAK are received along with their certificates
+            #  -Check if certificates are trusted
+            #  -Make sure the agent has generated the same keys as used in the certificates
+            if idevid_required:
+                if "iak_tpm" in json_body:
+                    iak_tpm = json_body["iak_tpm"]
+                    idevid_tpm = json_body["idevid_tpm"]
+                    iak_attest = base64.b64decode(json_body["iak_attest"])
+                    iak_sign = base64.b64decode(json_body["iak_sign"])
+                    logger.info("IDevID and IAK received")
+                    idevid_cert = json_body["idevid_cert"]
+                    iak_cert = json_body["iak_cert"]
+                else:
+                    logger.warning("Agent %s did not provide an IDevID/IAK", agent_id)
+                    web_util.echo_json_response(self, 400, "Error: no IDevID/IAK received")
+                    return
+
+                # IDevID and IAK cert checks
+                idevid_cert_509 = cert_utils.x509_der_cert(base64.b64decode(idevid_cert))
+                iak_cert_509 = cert_utils.x509_der_cert(base64.b64decode(iak_cert))
+                if not cert_utils.verify_cert(idevid_cert_509, config.get("tenant", "tpm_cert_store")):
+                    logger.warning("Invalid IDevID certificate for %s", agent_id)
+                    web_util.echo_json_response(self, 400, "Error: IDevID certificate could not be verified")
+                    return
+                logger.info("IDevID cert verified.")
+                if not cert_utils.verify_cert(iak_cert_509, config.get("tenant", "tpm_cert_store")):
+                    logger.warning("Invalid IAK certificate for %s", agent_id)
+                    web_util.echo_json_response(self, 400, "Error: IAK certificate could not be verified")
+                    return
+                logger.info("IAK cert verified.")
+                # NEEDS CRYPTO >= 38.0.0
+                # Check that the agent-generated keys match the certificates
```

Comment:
Could we not just send the IAK and IDevID certs from the agent? These seem like cross checks that also the agent could do if we don't trust the key-generation algorithm...

Reply:
Thanks for the comments on including cert types, I have pushed a commit with that included.

For this version we could only send the certs and take the keys from the certs. I did this here as I think there is a possibility we would want to allow sending the keys without certs from the agent and performing validation out-of-band and doing this here would reduce the amount of logic and config on the agent. 

Not especially attached to this though as that will require other changes in future as well, so happy to move these checks to the agent if you would prefer that. (This section will still require crypto 38.0.0 though)

Reply:
Also, you should check that the certificates we receive are indeed those of a TPM and not some other type of device that happens to have two certificates. The certificates have to validate against the certificate chain of the issuer and should distinguishable by the content of the SAN field where it indicates that it is a TPM (2.23.133.1.2).

Reply:
> Not especially attached to this though as that will require other changes in future as well, so happy to move these checks to the agent if you would prefer that. (This section will still require crypto 38.0.0 though)

I think the agent should report an error if it cannot properly recreate the key and bail out.

Reply:
> Also, you should check that the certificates we receive are indeed those of a TPM and not some other type of device that happens to have two certificates. The certificates have to validate against the certificate chain of the issuer and should distinguishable by the content of the SAN field where it indicates that it is a TPM (2.23.133.1.2).

I have added a check for this in [this commit](https://github.com/keylime/keylime/pull/1477/commits/69b08931c6b5631f38ffae0a2c766cd8ad81e88c)

Reply:
> I have added a check for this in [this commit](https://github.com/keylime/keylime/pull/1477/commits/69b08931c6b5631f38ffae0a2c766cd8ad81e88c)

Great!